{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Course in NEAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import random\n",
    "import pickle\n",
    "import copy\n",
    "import itertools\n",
    "import math\n",
    "import multiprocessing as mp\n",
    "import flappy_bird_gymnasium\n",
    "import gymnasium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utils:\n",
    "# Activation functions (used in forward pass, but not in the scope of the course)\n",
    "\n",
    "def sigmoid(x):\n",
    "    \"\"\"Return the S-Curve activation of x.\"\"\"\n",
    "    return 1/(1+math.exp(-x))\n",
    "\n",
    "def tanh(x):\n",
    "    \"\"\"Wrapper function for hyperbolic tangent activation.\"\"\"\n",
    "    return math.tanh(x)\n",
    "\n",
    "def LReLU(x):\n",
    "    \"\"\"Leaky ReLU function for x\"\"\"\n",
    "    if x >= 0:\n",
    "        return x\n",
    "    else:\n",
    "        return 0.01 * x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Hyperparameters**\n",
    "\n",
    "- delta_threshold - (?) The threshold used to determine wether an agent belongs to a specific  specie\n",
    "- distance_weights - The weights used to determine the genomic distance between geneotypes, used in speciation.\n",
    "        - edge: the synapse/connection between two neurons\n",
    "        - weight: the value given to the edge\n",
    "        - bias: the bias value of the node\n",
    "- default_activation - the activation function used in forward pass. In this course we can choose between LReLU, tanh and sigmoid. \n",
    "- max_fitness - Stop condition. The desiered fitness value that stops the evolution when reached\n",
    "- max_generations - Stop condition. The number of generations before the evolution stops if max_fitness is not reached.\n",
    "- max_fitness_history: (?) The number of max fitness for a generation that is kept track on.\n",
    "- breed_probabilities - We have sexual and asexual crossover/breeding in this algorithm, and there is a random chance which one is chosen.\n",
    "- mutation_probabilities - the probabilities for each type of mutation\n",
    "            'node' : add or remove a node,\n",
    "            'edge' : add or remove an edge,\n",
    "            'weight_perturb' : perturburate a weight,\n",
    "            'weight_set' : set a new value to a weight,\n",
    "            'bias_perturb' : perturburate a bias,\n",
    "            'bias_set: set a new value to a bias.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Hyperparameters(object):\n",
    "    \"\"\"Hyperparameter settings.\"\"\"\n",
    "    def __init__(self):\n",
    "        self.delta_threshold = 1.5\n",
    "        self.distance_weights = {\n",
    "            'edge' : 1.0,\n",
    "            'weight' : 1.0,\n",
    "            'bias' : 1.0\n",
    "        }\n",
    "        self.default_activation = sigmoid\n",
    "\n",
    "        self.max_fitness = float('inf')\n",
    "        self.max_generations = float('inf')\n",
    "        self.max_fitness_history = 30\n",
    "\n",
    "        self.breed_probabilities = {\n",
    "            'asexual' : 0.5,\n",
    "            'sexual' : 0.5\n",
    "        }\n",
    "        self.mutation_probabilities = {\n",
    "            'node' : 0.01,\n",
    "            'edge' : 0.09,\n",
    "            'weight_perturb' : 0.4,\n",
    "            'weight_set' : 0.1,\n",
    "            'bias_perturb' : 0.3,\n",
    "            'bias_set' : 0.1\n",
    "        }\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Defining the neural network**\n",
    "The neural network of each agent will be defined by the genome if the individual. The genes will consist of data for nodes and edges. therefore we need to establish an Edge class and a Node class that keeps track on the weighs and biases, activation, output and whether the edge is enabled. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Edge(object):\n",
    "    \"\"\"A gene object representing an edge in the neural network.\"\"\"\n",
    "    def __init__(self, weight):\n",
    "        self.weight = weight\n",
    "        self.enabled = True\n",
    "\n",
    "\n",
    "class Node(object):\n",
    "    \"\"\"A gene object representing a node in the neural network.\"\"\"\n",
    "    def __init__(self, activation):\n",
    "        self.output = 0\n",
    "        self.bias = 0\n",
    "        self.activation = activation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**The Genome**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Genome(object):\n",
    "    \"\"\"Base class for a standard genome used by the NEAT algorithm.\"\"\"\n",
    "    def __init__(self, inputs, outputs, default_activation):\n",
    "        # Nodes\n",
    "        self._inputs = inputs\n",
    "        self._outputs = outputs\n",
    "\n",
    "        self._unhidden = inputs+outputs\n",
    "        self._max_node = inputs+outputs\n",
    "\n",
    "        # Structure\n",
    "        self._edges = {} # (i, j) : Edge\n",
    "        self._nodes = {} # NodeID : Node\n",
    "\n",
    "        self._default_activation = default_activation\n",
    "\n",
    "        # Performance\n",
    "        self._fitness = 0\n",
    "        self._adjusted_fitness = 0\n",
    "\n",
    "    def generate(self):\n",
    "        \"\"\"Generate the neural network of this genome with minimal\n",
    "        initial topology, i.e. (no hidden nodes). Call on genome\n",
    "        creation.\n",
    "        \"\"\"\n",
    "        # Minimum initial topology, no hidden layer\n",
    "        for n in range(self._max_node):\n",
    "            self._nodes[n] = Node(self._default_activation)\n",
    "\n",
    "        for i in range(self._inputs):\n",
    "            for j in range(self._inputs, self._unhidden):\n",
    "                self.add_edge(i, j, random.uniform(-1, 1))\n",
    "                \n",
    "    def forward(self, inputs):\n",
    "        \"\"\"Evaluate inputs and calculate the outputs of the\n",
    "        neural network via the forward propagation algorithm.\n",
    "        \"\"\"\n",
    "        if len(inputs) != self._inputs:\n",
    "            raise ValueError(\"Incorrect number of inputs.\")\n",
    "\n",
    "        # Set input values\n",
    "        for i in range(self._inputs):\n",
    "            self._nodes[i].output = inputs[i]\n",
    "        \n",
    "        # Generate backward-adjacency list \n",
    "        _from = {}\n",
    "        for n in range(self._max_node):\n",
    "            _from[n] = []\n",
    "\n",
    "        for (i, j) in self._edges:\n",
    "            if not self._edges[(i, j)].enabled:\n",
    "                continue\n",
    "            _from[j].append(i)\n",
    "\n",
    "        # Calculate output values for each node\n",
    "        ordered_nodes = itertools.chain(\n",
    "            range(self._unhidden, self._max_node),\n",
    "            range(self._inputs, self._unhidden)\n",
    "        )\n",
    "        for j in ordered_nodes:\n",
    "            ax = 0\n",
    "            for i in _from[j]:\n",
    "                ax += self._edges[(i, j)].weight * self._nodes[i].output\n",
    "\n",
    "            node = self._nodes[j]\n",
    "            node.output = node.activation(ax + node.bias)\n",
    "        \n",
    "        return [self._nodes[n].output for n in range(self._inputs, self._unhidden)]\n",
    "\n",
    "    def mutate(self, probabilities):\n",
    "        \"\"\"Randomly mutate the genome to initiate variation.\"\"\"\n",
    "        if self.is_disabled():\n",
    "            self.add_enabled()\n",
    "\n",
    "        population = list(probabilities.keys())\n",
    "        weights = [probabilities[k] for k in population]\n",
    "        choice = random.choices(population, weights=weights)[0]\n",
    "\n",
    "        if choice == \"node\":\n",
    "            self.add_node()\n",
    "        elif choice == \"edge\":\n",
    "            (i, j) = self.random_pair()\n",
    "            self.add_edge(i, j, random.uniform(-1, 1))\n",
    "        elif choice == \"weight_perturb\" or choice == \"weight_set\":\n",
    "            self.shift_weight(choice)\n",
    "        elif choice == \"bias_perturb\" or choice == \"bias_set\":\n",
    "            self.shift_bias(choice)\n",
    "\n",
    "        self.reset()\n",
    "\n",
    "    def add_node(self):\n",
    "        \"\"\"Add a new node between a randomly selected edge,\n",
    "        disabling the parent edge.\n",
    "        \"\"\"\n",
    "        enabled = [k for k in self._edges if self._edges[k].enabled]\n",
    "        (i, j) = random.choice(enabled)\n",
    "        edge = self._edges[(i, j)]\n",
    "        edge.enabled = False\n",
    "\n",
    "        new_node = self._max_node\n",
    "        self._max_node += 1\n",
    "        self._nodes[new_node] = Node(self._default_activation)\n",
    "\n",
    "        self.add_edge(i, new_node, 1.0)\n",
    "        self.add_edge(new_node, j, edge.weight)\n",
    "\n",
    "    def add_edge(self, i, j, weight):\n",
    "        \"\"\"Add a new connection between existing nodes.\"\"\"\n",
    "        if (i, j) in self._edges:\n",
    "            self._edges[(i, j)].enabled = True\n",
    "        else:\n",
    "            self._edges[(i, j)] = Edge(weight)\n",
    "            \n",
    "    def add_enabled(self):\n",
    "        \"\"\"Re-enable a random disabled edge.\"\"\"\n",
    "        disabled = [e for e in self._edges if not self._edges[e].enabled]\n",
    "\n",
    "        if len(disabled) > 0:\n",
    "            self._edges[random.choice(disabled)].enabled = True\n",
    "        \n",
    "    def shift_weight(self, type):\n",
    "        \"\"\"Randomly shift, perturb, or set one of the edge weights.\"\"\"\n",
    "        e = random.choice(list(self._edges.keys()))\n",
    "        if type == \"weight_perturb\":\n",
    "            self._edges[e].weight += random.uniform(-1, 1)\n",
    "        elif type == \"weight_set\":\n",
    "            self._edges[e].weight = random.uniform(-1, 1)\n",
    "\n",
    "    def shift_bias(self, type):\n",
    "        \"\"\"Randomly shift, perturb, or set the bias of an incoming edge.\"\"\"\n",
    "        # Select only nodes in the hidden and output layer\n",
    "        n = random.choice(range(self._inputs, self._max_node))\n",
    "        if type == \"bias_perturb\":\n",
    "            self._nodes[n].bias += random.uniform(-1, 1)\n",
    "        elif type == \"bias_set\":\n",
    "            self._nodes[n].bias = random.uniform(-1, 1)\n",
    "\n",
    "    def random_pair(self):\n",
    "        \"\"\"Generate random nodes (i, j) such that:\n",
    "        1. i is not an output\n",
    "        2. j is not an input\n",
    "        3. i != j\n",
    "        \"\"\"\n",
    "        i = random.choice([n for n in range(self._max_node) if not self.is_output(n)])\n",
    "        j_list = [n for n in range(self._max_node) if not self.is_input(n) and n != i]\n",
    "\n",
    "        if not j_list:\n",
    "            j = self._max_node\n",
    "            self.add_node()\n",
    "        else:\n",
    "            j = random.choice(j_list)\n",
    "\n",
    "        return (i, j)\n",
    "\n",
    "    def is_input(self, n):\n",
    "        \"\"\"Determine if the node id is an input.\"\"\"\n",
    "        return 0 <= n < self._inputs\n",
    "\n",
    "    def is_output(self, n):\n",
    "        \"\"\"Determine if the node id is an output.\"\"\"\n",
    "        return self._inputs <= n < self._unhidden\n",
    "\n",
    "    def is_hidden(self, n):\n",
    "        \"\"\"Determine if the node id is hidden.\"\"\"\n",
    "        return self._unhidden <= n < self._max_node\n",
    "\n",
    "    def is_disabled(self):\n",
    "        \"\"\"Determine if all of its genes are disabled.\"\"\"\n",
    "        return all(self._edges[i].enabled == False for i in self._edges)\n",
    "\n",
    "    def get_fitness(self):\n",
    "        \"\"\"Return the fitness of the genome.\"\"\"\n",
    "        return self._fitness\n",
    "\n",
    "    def get_nodes(self):\n",
    "        \"\"\"Get the nodes of the network.\"\"\"\n",
    "        return self._nodes.copy()\n",
    "\n",
    "    def get_edges(self):\n",
    "        \"\"\"Get the network's edges.\"\"\"\n",
    "        return self._edges.copy()\n",
    "\n",
    "    def get_num_nodes(self):\n",
    "        \"\"\"Get the number of nodes in the network.\"\"\"\n",
    "        return self._max_node\n",
    "\n",
    "    def set_fitness(self, score):\n",
    "        \"\"\"Set the fitness score of this genome.\"\"\"\n",
    "        self._fitness = score\n",
    "\n",
    "    def reset(self):\n",
    "        \"\"\"Reset the genome's internal state.\"\"\"\n",
    "        for n in range(self._max_node):\n",
    "            self._nodes[n].output = 0\n",
    "        self._fitness = 0\n",
    "\n",
    "    def clone(self):\n",
    "        \"\"\"Return a clone of the genome.\n",
    "        \"\"\"\n",
    "        return copy.deepcopy(self)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genomic_distance(a, b, distance_weights):\n",
    "    \"\"\"Calculate the genomic distance between two genomes.\"\"\"\n",
    "    a_edges = set(a._edges)\n",
    "    b_edges = set(b._edges)\n",
    "\n",
    "    # Does not distinguish between disjoint and excess\n",
    "    matching_edges = a_edges & b_edges\n",
    "    disjoint_edges = (a_edges - b_edges) | (b_edges - a_edges)\n",
    "    N_edges = len(max(a_edges, b_edges, key=len))\n",
    "    N_nodes = min(a._max_node, b._max_node)\n",
    "\n",
    "    weight_diff = 0\n",
    "    for i in matching_edges:\n",
    "        weight_diff += abs(a._edges[i].weight - b._edges[i].weight)\n",
    "\n",
    "    bias_diff = 0\n",
    "    for i in range(N_nodes):\n",
    "        bias_diff += abs(a._nodes[i].bias - b._nodes[i].bias)\n",
    "\n",
    "    t1 = distance_weights['edge'] * len(disjoint_edges)/N_edges\n",
    "    t2 = distance_weights['weight'] * weight_diff/len(matching_edges)\n",
    "    t3 = distance_weights['bias'] * bias_diff/N_nodes\n",
    "    return t1 + t2 + t3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genomic_crossover(a, b):\n",
    "    \"\"\"Breed two genomes and return the child. Matching genes\n",
    "    are inherited randomly, while disjoint genes are inherited\n",
    "    from the fitter parent.\n",
    "    \"\"\"\n",
    "    # Template genome for child\n",
    "    child = Genome(a._inputs, a._outputs, a._default_activation)\n",
    "    a_in = set(a._edges)\n",
    "    b_in = set(b._edges)\n",
    "\n",
    "    # Inherit homologous gene from a random parent\n",
    "    for i in a_in & b_in:\n",
    "        parent = random.choice([a, b])\n",
    "        child._edges[i] = copy.deepcopy(parent._edges[i])\n",
    "\n",
    "    # Inherit disjoint/excess genes from fitter parent\n",
    "    if a._fitness > b._fitness:\n",
    "        for i in a_in - b_in:\n",
    "            child._edges[i] = copy.deepcopy(a._edges[i])\n",
    "    else:\n",
    "        for i in b_in - a_in:\n",
    "            child._edges[i] = copy.deepcopy(b._edges[i])\n",
    "    \n",
    "    # Calculate max node\n",
    "    child._max_node = 0\n",
    "    for (i, j) in child._edges:\n",
    "        current_max = max(i, j)\n",
    "        child._max_node = max(child._max_node, current_max)\n",
    "    child._max_node += 1\n",
    "\n",
    "    # Inherit nodes\n",
    "    for n in range(child._max_node):\n",
    "        inherit_from = []\n",
    "        if n in a._nodes:\n",
    "            inherit_from.append(a)\n",
    "        if n in b._nodes:\n",
    "            inherit_from.append(b)\n",
    "\n",
    "        random.shuffle(inherit_from)\n",
    "        parent = max(inherit_from, key=lambda p: p._fitness)\n",
    "        child._nodes[n] = copy.deepcopy(parent._nodes[n])\n",
    "\n",
    "    child.reset()\n",
    "    return child"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Specie**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Specie(object):\n",
    "    \"\"\"A specie represents a set of genomes whose genomic distances \n",
    "    between them fall under the Brain's delta threshold.\n",
    "    \"\"\"\n",
    "    def __init__(self, max_fitness_history, *members):\n",
    "        self._members = list(members)\n",
    "        self._fitness_history = []\n",
    "        self._fitness_sum = 0\n",
    "        self._max_fitness_history = max_fitness_history\n",
    "\n",
    "    def breed(self, mutation_probabilities, breed_probabilities):\n",
    "        \"\"\"Return a child as a result of either a mutated clone\n",
    "        or crossover between two parent genomes.\n",
    "        \"\"\"\n",
    "        # Either mutate a clone or breed two random genomes\n",
    "        population = list(breed_probabilities.keys())\n",
    "        probabilities= [breed_probabilities[k] for k in population]\n",
    "        choice = random.choices(population, weights=probabilities)[0]\n",
    "\n",
    "        if choice == \"asexual\" or len(self._members) == 1:\n",
    "            child = random.choice(self._members).clone()\n",
    "            child.mutate(mutation_probabilities)\n",
    "        elif choice == \"sexual\":\n",
    "            (mom, dad) = random.sample(self._members, 2)\n",
    "            child = genomic_crossover(mom, dad)\n",
    "\n",
    "        return child\n",
    "\n",
    "    def update_fitness(self):\n",
    "        \"\"\"Update the adjusted fitness values of each genome \n",
    "        and the historical fitness.\"\"\"\n",
    "        for g in self._members:\n",
    "            g._adjusted_fitness = g._fitness/len(self._members)\n",
    "\n",
    "        self._fitness_sum = sum([g._adjusted_fitness for g in self._members])\n",
    "        self._fitness_history.append(self._fitness_sum)\n",
    "        if len(self._fitness_history) > self._max_fitness_history:\n",
    "            self._fitness_history.pop(0)\n",
    "\n",
    "    def cull_genomes(self, fittest_only):\n",
    "        \"\"\"Exterminate the weakest genomes per specie.\"\"\"\n",
    "        self._members.sort(key=lambda g: g._fitness, reverse=True)\n",
    "        if fittest_only:\n",
    "            # Only keep the winning genome\n",
    "            remaining = 1\n",
    "        else:\n",
    "            # Keep top 25%\n",
    "            remaining = int(math.ceil(0.25*len(self._members)))\n",
    "\n",
    "        self._members = self._members[:remaining]\n",
    "\n",
    "    def get_best(self):\n",
    "        \"\"\"Get the member with the highest fitness score.\"\"\"\n",
    "        return max(self._members, key=lambda g: g._fitness)\n",
    "\n",
    "    def can_progress(self):\n",
    "        \"\"\"Determine whether species should survive the culling.\"\"\"\n",
    "        n = len(self._fitness_history)\n",
    "        avg = sum(self._fitness_history) / n\n",
    "        return avg > self._fitness_history[0] or n < self._max_fitness_history\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Evolution**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Brain(object):\n",
    "    \"\"\"Base class for a 'brain' that learns through the evolution\n",
    "    of a population of genomes.\n",
    "    \"\"\"\n",
    "    def __init__(self, inputs, outputs, population=100, hyperparams=Hyperparameters()):\n",
    "        self._inputs = inputs\n",
    "        self._outputs = outputs\n",
    "\n",
    "        self._species = []\n",
    "        self._population = population\n",
    "\n",
    "        # Hyper-parameters\n",
    "        self._hyperparams = hyperparams\n",
    "        \n",
    "        self._generation = 0\n",
    "        self._current_species = 0\n",
    "        self._current_genome = 0\n",
    "\n",
    "        self._global_best = None\n",
    "\n",
    "    def generate(self):\n",
    "        \"\"\"Generate the initial population of genomes.\"\"\"\n",
    "        for i in range(self._population):\n",
    "            g = Genome(self._inputs, self._outputs, \n",
    "                       self._hyperparams.default_activation)\n",
    "            g.generate()\n",
    "            self.classify_genome(g)\n",
    "        \n",
    "        # Set the initial best genome\n",
    "        self._global_best = self._species[0]._members[0]\n",
    "\n",
    "    def classify_genome(self, genome):\n",
    "        \"\"\"Classify genomes into species via the genomic\n",
    "        distance algorithm.\n",
    "        \"\"\"\n",
    "        if not self._species:\n",
    "            # Empty population\n",
    "            self._species.append(Specie(\n",
    "                    self._hyperparams.max_fitness_history, genome\n",
    "                )\n",
    "            )\n",
    "        else:\n",
    "            # Compare genome against representative s[0] in each specie\n",
    "            for s in self._species:\n",
    "                rep =  s._members[0]\n",
    "                dist = genomic_distance(\n",
    "                    genome, rep, self._hyperparams.distance_weights\n",
    "                )\n",
    "                if dist <= self._hyperparams.delta_threshold:\n",
    "                    s._members.append(genome)\n",
    "                    return\n",
    "\n",
    "            # Doesn't fit with any other specie, create a new one\n",
    "            self._species.append(Specie(\n",
    "                    self._hyperparams.max_fitness_history, genome\n",
    "                )\n",
    "            )\n",
    "\n",
    "    def update_fittest(self):\n",
    "        \"\"\"Update the highest fitness score of the whole population.\"\"\"\n",
    "        top_performers = [s.get_best() for s in self._species]\n",
    "        current_top = max(top_performers, key=lambda g: g._fitness)\n",
    "\n",
    "        if current_top._fitness > self._global_best._fitness:\n",
    "            self._global_best = current_top.clone()\n",
    "\n",
    "    def evolve(self):\n",
    "        \"\"\"Evolve the population by eliminating the poorest performing\n",
    "        genomes and repopulating with mutated children, prioritizing\n",
    "        the most promising species.\n",
    "        \"\"\"\n",
    "        global_fitness_sum = 0\n",
    "        for s in self._species:\n",
    "            s.update_fitness()\n",
    "            global_fitness_sum += s._fitness_sum\n",
    "\n",
    "        if global_fitness_sum == 0:\n",
    "            # No progress, mutate everybody\n",
    "            for s in self._species:\n",
    "                for g in s._members:\n",
    "                    g.mutate(self._hyperparams.mutation_probabilities)\n",
    "        else:\n",
    "            # Only keep the species with potential to improve\n",
    "            surviving_species = []\n",
    "            for s in self._species:\n",
    "                if s.can_progress():\n",
    "                    surviving_species.append(s)\n",
    "            self._species = surviving_species\n",
    "\n",
    "            # Eliminate lowest performing genomes per specie\n",
    "            for s in self._species:\n",
    "                s.cull_genomes(False)\n",
    "\n",
    "            # Repopulate\n",
    "            for i, s in enumerate(self._species):\n",
    "                ratio = s._fitness_sum/global_fitness_sum\n",
    "                diff = self._population - self.get_population()\n",
    "                offspring = int(round(ratio * diff))\n",
    "                for j in range(offspring):\n",
    "                    self.classify_genome(\n",
    "                        s.breed(\n",
    "                            self._hyperparams.mutation_probabilities, \n",
    "                            self._hyperparams.breed_probabilities\n",
    "                        )\n",
    "                    )\n",
    "\n",
    "            # No species survived\n",
    "            # Repopulate using mutated minimal structures and global best\n",
    "            if not self._species:\n",
    "                for i in range(self._population):\n",
    "                    if i%3 == 0:\n",
    "                        g = self._global_best.clone()\n",
    "                    else:\n",
    "                        g = Genome(self._inputs, self._outputs, \n",
    "                                   self._hyperparams.default_activation)\n",
    "                        g.generate()\n",
    "                    g.mutate(self._hyperparams.mutation_probabilities)\n",
    "                    self.classify_genome(g)\n",
    "\n",
    "        self._generation += 1\n",
    "\n",
    "    def should_evolve(self):\n",
    "        \"\"\"Determine if the system should continue to evolve\n",
    "        based on the maximum fitness and generation count.\n",
    "        \"\"\"\n",
    "        self.update_fittest()\n",
    "        fit = self._global_best._fitness <= self._hyperparams.max_fitness\n",
    "        end = self._generation != self._hyperparams.max_generations\n",
    "\n",
    "        return fit and end\n",
    "\n",
    "    def next_iteration(self):\n",
    "        \"\"\"Call after every evaluation of individual genomes to\n",
    "        progress training.\n",
    "        \"\"\"\n",
    "        s = self._species[self._current_species]\n",
    "        if self._current_genome < len(s._members)-1:\n",
    "            self._current_genome += 1\n",
    "        else:\n",
    "            if self._current_species < len(self._species)-1:\n",
    "                self._current_species += 1\n",
    "                self._current_genome = 0\n",
    "            else:\n",
    "                # Evolve to the next generation\n",
    "                self.evolve()\n",
    "                self._current_species = 0\n",
    "                self._current_genome = 0\n",
    "\n",
    "    def evaluate_parallel(self, evaluator, *args, **kwargs):\n",
    "        \"\"\"Evaluate the entire population on separate processes\n",
    "        to progress training. The evaluator function must take a Genome\n",
    "        as its first parameter and return a numerical fitness score.\n",
    "\n",
    "        Any global state passed to the evaluator is copied and will not\n",
    "        be modified at the parent process.\n",
    "        \"\"\"\n",
    "        max_proc = max(mp.cpu_count()-1, 1)\n",
    "        pool = mp.Pool(processes=max_proc)\n",
    "        \n",
    "        results = {}\n",
    "        for i in range(len(self._species)):\n",
    "            for j in range(len(self._species[i]._members)):\n",
    "                results[(i, j)] = pool.apply_async(\n",
    "                    evaluator, \n",
    "                    args=[self._species[i]._members[j]]+list(args), \n",
    "                    kwds=kwargs\n",
    "                )\n",
    "\n",
    "        for key in results:\n",
    "            genome = self._species[key[0]]._members[key[1]]\n",
    "            genome.set_fitness(results[key].get())\n",
    "\n",
    "        pool.close()\n",
    "        pool.join()\n",
    "        self.evolve()\n",
    "\n",
    "    def get_fittest(self):\n",
    "        \"\"\"Return the genome with the highest global fitness score.\"\"\"\n",
    "        return self._global_best\n",
    "\n",
    "    def get_population(self):\n",
    "        \"\"\"Return the true (calculated) population size.\"\"\"\n",
    "        return sum([len(s._members) for s in self._species])\n",
    "\n",
    "    def get_current(self):\n",
    "        \"\"\"Get the current genome for evaluation.\"\"\"\n",
    "        s = self._species[self._current_species]\n",
    "        return s._members[self._current_genome]\n",
    "\n",
    "    def get_current_species(self):\n",
    "        \"\"\"Get index of current species being evaluated.\"\"\"\n",
    "        return self._current_species\n",
    "\n",
    "    def get_current_genome(self):\n",
    "        \"\"\"Get index of current genome being evaluated.\"\"\"\n",
    "        return self._current_genome\n",
    "\n",
    "    def get_generation(self):\n",
    "        \"\"\"Get the current generation number of this population.\"\"\"\n",
    "        return self._generation\n",
    "\n",
    "    def get_species(self):\n",
    "        \"\"\"Get the list of species and their respective member genomes.\"\"\"\n",
    "        return self._species\n",
    "\n",
    "    def save(self, filename):\n",
    "        \"\"\"Save an instance of the population to disk.\"\"\"\n",
    "        with open(filename+'.neat', 'wb') as _out:\n",
    "            pickle.dump(self, _out, pickle.HIGHEST_PROTOCOL)\n",
    "\n",
    "    @staticmethod\n",
    "    def load(filename):\n",
    "        \"\"\"Return an instance of a population from disk.\"\"\"\n",
    "        with open(filename+'.neat', 'rb') as _in:\n",
    "            return pickle.load(_in)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Simulating evolution**\n",
    "The most tedious stage of evolutionary algorithms is usually the simulation stage where the population of agents tries to solve the task, and gets an fitness evaluation. This is normally done in a game loop that simulates an environment. Environments can differ from everything from a game environment to a set of mathematical rules where the neural networks makes predictions. In this case we are using flappy bird."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 0.98611111  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.4609375  -0.8         0.46666667]\n",
      "[ 0.97222222  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.44726562 -0.7         0.43333333]\n",
      "[ 0.95833333  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.43554688 -0.6         0.4       ]\n",
      "[ 0.94444444  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.42578125 -0.5         0.36666667]\n",
      "[ 0.93055556  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.41796875 -0.4         0.33333333]\n",
      "[ 0.91666667  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.40039062 -0.9         0.5       ]\n",
      "[ 0.90277778  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.3828125  -0.9         0.5       ]\n",
      "[ 0.88888889  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.3671875  -0.8         0.46666667]\n",
      "[ 0.875       0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.35351562 -0.7         0.43333333]\n",
      "[ 0.86111111  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.34179688 -0.6         0.4       ]\n",
      "[ 0.84722222  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.33203125 -0.5         0.36666667]\n",
      "[ 0.83333333  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.32421875 -0.4         0.33333333]\n",
      "[ 0.81944444  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.31835938 -0.3         0.3       ]\n",
      "[ 0.80555556  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.30078125 -0.9         0.5       ]\n",
      "[ 0.79166667  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.28515625 -0.8         0.46666667]\n",
      "[ 0.77777778  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.27148438 -0.7         0.43333333]\n",
      "[ 0.76388889  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.25976562 -0.6         0.4       ]\n",
      "[ 0.75       0.2734375  0.46875    1.         0.         1.\n",
      "  1.         0.         1.         0.2421875 -0.9        0.5      ]\n",
      "[ 0.73611111  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.2265625  -0.8         0.46666667]\n",
      "[ 0.72222222  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.21289062 -0.7         0.43333333]\n",
      "[ 0.70833333  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.20117188 -0.6         0.4       ]\n",
      "[ 0.69444444  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.19140625 -0.5         0.36666667]\n",
      "[ 0.68055556  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.17382812 -0.9         0.5       ]\n",
      "[ 0.66666667  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.15820312 -0.8         0.46666667]\n",
      "[ 0.65277778  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.140625   -0.9         0.5       ]\n",
      "[ 0.63888889  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.125      -0.8         0.46666667]\n",
      "[ 0.625       0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.11132812 -0.7         0.43333333]\n",
      "[ 0.61111111  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.09960938 -0.6         0.4       ]\n",
      "[ 0.59722222  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.08203125 -0.9         0.5       ]\n",
      "[ 0.58333333  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.06445312 -0.9         0.5       ]\n",
      "[ 0.56944444  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.04882812 -0.8         0.46666667]\n",
      "[ 0.55555556  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.03125    -0.9         0.5       ]\n",
      "[ 0.54166667  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.015625   -0.8         0.46666667]\n",
      "[ 0.52777778  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.          0.00195312 -0.7         0.43333333]\n",
      "[ 0.51388889  0.2734375   0.46875     1.          0.          1.\n",
      "  1.          0.          1.         -0.00976562 -0.6         0.4       ]\n",
      "[ 0.5         0.2734375   0.46875     1.          0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.02734375 -0.9         0.5       ]\n",
      "[ 0.48611111  0.2734375   0.46875     0.98611111  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.04296875 -0.8         0.46666667]\n",
      "[ 0.47222222  0.2734375   0.46875     0.97222222  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.05664062 -0.7         0.43333333]\n",
      "[ 0.45833333  0.2734375   0.46875     0.95833333  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.07421875 -0.9         0.5       ]\n",
      "[ 0.44444444  0.2734375   0.46875     0.94444444  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.08984375 -0.8         0.46666667]\n",
      "[ 0.43055556  0.2734375   0.46875     0.93055556  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.10351562 -0.7         0.43333333]\n",
      "[ 0.41666667  0.2734375   0.46875     0.91666667  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.11523438 -0.6         0.4       ]\n",
      "[ 0.40277778  0.2734375   0.46875     0.90277778  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.125      -0.5         0.36666667]\n",
      "[ 0.38888889  0.2734375   0.46875     0.88888889  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.1328125  -0.4         0.33333333]\n",
      "[ 0.375       0.2734375   0.46875     0.875       0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.13867188 -0.3         0.3       ]\n",
      "[ 0.36111111  0.2734375   0.46875     0.86111111  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.14257812 -0.2         0.26666667]\n",
      "[ 0.34722222  0.2734375   0.46875     0.84722222  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.14453125 -0.1         0.23333333]\n",
      "[ 0.33333333  0.2734375   0.46875     0.83333333  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.14453125  0.          0.2       ]\n",
      "[ 0.31944444  0.2734375   0.46875     0.81944444  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.14257812  0.1         0.16666667]\n",
      "[ 0.30555556  0.2734375   0.46875     0.80555556  0.33203125  0.52734375\n",
      "  1.          0.          1.         -0.13867188  0.2         0.13333333]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "env = gymnasium.make(\"FlappyBird-v0\", render_mode=\"human\", use_lidar=False)\n",
    "\n",
    "\n",
    "# ===============================================================\n",
    "# Neural network parameters:\n",
    "# ===============================================================\n",
    "num_inputs = 12 # The envirement has 12 observations for each frame\n",
    "num_outputs = 1 # The envirement has 1 action space (flap or do nothing)\n",
    "population_size = 100\n",
    "hyperparameters = Hyperparameters()\n",
    "    \n",
    "    \n",
    "obs, _ = env.reset()\n",
    "while True:\n",
    "    \n",
    "    \n",
    "    \n",
    "    # ===============================================================\n",
    "    # Next action:\n",
    "    # ===============================================================\n",
    "    # (feed the observation to your agent here)\n",
    "    # Action space:\n",
    "    # 0 - do nothing\n",
    "    # 1 - flap\n",
    "    #action = env.action_space.sample()\n",
    "    action = random.randint(0,1)\n",
    "\n",
    "    # Processing:\n",
    "    obs, reward, terminated, _, info = env.step(action)\n",
    "    print(obs)\n",
    "    # Rewards:\n",
    "    # +0.1 - every frame it stays alive\n",
    "    # +1.0 - successfully passing a pipe\n",
    "    # -1.0 - dying\n",
    "    # −0.5 - touch the top of the screen\n",
    "\n",
    "    \n",
    "    # Checking if the player is still alive\n",
    "    if terminated:\n",
    "        break\n",
    "    \n",
    "  \n",
    "\n",
    "env.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
